### ISSUE TITLE
Prolific recruiter does not listen for returned or abandoned notifications

### ISSUE THREAD

**Issue by thomasmorgan** [Dallinger/Dallinger#5725]
The prolific recruiter listens for submitted notifications, but not returned or abandoned notifications. This is sort of a missing feature as opposed to a bug, but the functionality is so critical I've labeled it as a bug anyway.

## Expected Behavior
The prolific recruiter should listen for returned and abandoned notifications and process them accordingly. This functionality is already present for the MTurk recruiter.

## Current Behavior
If a participant returns a hit, Dallinger cannot fail their nodes etc. because it is not processing any notification that the participant returned the hit. However, prolific still recruits a replacement. This can cause participants to arrive at inappropriate times, for instance where there is no space in the networks for them. in such cases, participants often return the hit (resulting in another participant being recruited by prolific) or submit an incomplete hit (resulting in Dallinger auto_recruiting a replacement due to the incomplete submission being picked up by data_check). In both cases this can lead to a recruitment spiral which wastes money.

## Steps to Reproduce (for bugs)
Run any experiment on prolific. Wait until a participant returns the hit part way through an experiment. Note that the participants nodes, etc. are not failed.

You can also just look at the code for the prolific recruiter - there is code for submitted notifications, but no others.

## Context
We were running an experiment in which networks consisted of pairs of nodes, each associated with a different participant. When one participant returns, we need to replace them. Prolific does automatically send in a replacement. However, without a notification to indicate which participant needs to be replaced it is not possible to add the new participant to the network and the experiment stalls.

## Your Environment
* Dallinger version used: Most recent - can't remember off the top of my head, sorry!
* Operating System and version (desktop or mobile): Linux, desktop
* Browser and version (e.g. Chrome 39, node.js 5.4): Not relevant
* Link to your project: https://github.com/Ruben222/Epistemic-Vigilanceexpt

**Comment by jessesnyder** [comment]
https://docs.prolific.com/docs/api-docs/public/#tag/Submissions/Submissions-guide

**Comment by jessesnyder** [comment]
@thomasmorgan @jacobyn Prolific does not send notifications, so we will need to poll, based on a clock task. We already use the clock to tell recruiters about overdue participants based on `config.get("duration")`, but the ProlificRecruiter doesn't do anything in response currently. This should be a reasonably straightforward enhancement. However, it sounds like you also need to know about _returned_ participants more or less as they happen (and definitely prior to exceeding the experiment duration). Is that correct?

**Comment by jessesnyder** [comment]
Also, the MTurkRecruiter does a lot of [fancy stuff ](https://github.com/Dallinger/Dallinger/blob/263328179d41aaf115020dc04c61ee9497bd5d0e/dallinger/recruiters.py#L1154)when experiment duration is exceeded (emails to the researcher, attempting to close the HIT, etc.) Do we want to emulate that for Prolific?

**Comment by pmcharrison** [comment]
@polvanrijn has some thoughts about this too. I think that a clock process would work well, though I think the actual queries should happen in a worker process to prevent blocking (unless you can use some async logic within the clock process itself?)

**Comment by polvanrijn** [comment]
Yes, we were thinking it could be handy to periodically poll the recruiter. From polling the recruiter, we will get quite some stats (e.g., total cost of experiment) but also the 'true' recruiter status of participants. This will also be very handy to determine how many participants returned or to find out how many 'alive' participants there are in the experiment. Especially the last point is often hard to find out and can be quite relevant depending on the experiment (e.g. for synchronous experiments). Also, this allows us to see any potential discrepancies in paid bonus + base pay to our internally computed numbers.

**Comment by pmcharrison** [comment]
Yep, I think some general process like @polvanrijn outlines would be very good for making sure all of our numbers stay up-to-date and consistent. We have recently talked about refactoring recruiter classes into a common interface; it might be wise to do that refactoring prior to the present issue, that way we can get this functionality into all the recruiters for the price of one.

**Comment by pmcharrison** [comment]
Thanks @thomasmorgan!

I agree with not replicating the MTurk 'fancy stuff' (emails to the researcher, attempting to close the HIT, etc.) with Prolific, sorry for missing that question before.

> Rather than tying this solely to a clock process (creating a race between the dallinger process and prolific that dallinger might lose), it may be more robust to confirm all existing participant statuses whenever a new participant arrives. That way all returned participants will be detected before the new participant has a node created for them.

If we put this in the clock process we would set a background task that runs every e.g. 30 seconds and identifies any returned participants. We could then make this process call `recruit()` if required. Then I think there would be no race condition? I think this would be more efficient than waiting for the next participant to arrive before checking.

**Comment by jessesnyder** [comment]
It looks like they're working on an event subscription model (I don't remember seeing this before):
https://docs.prolific.com/docs/api-docs/public/#tag/Hooks

However, note the immediate disclaimer: "This is an experimental feature that may be subject to change in the future."

It looks like the support person I was working with (Elaine Poon) when developing the ProlificRecruiter isn't around anymore. Do we have a new technical support contact we could ask about the actual maturity/stability of this API?

**Comment by pmcharrison** [comment]
I'm not sure about the Prolific support contact, but maybe @jacobyn knows?

**Comment by pmcharrison** [comment]
There is something reassuring about the simple polling approach though, in that you know you have a fully complete record; otherwise a missed notification can mess things up.

**Comment by pmcharrison** [comment]
Hi @thomasmorgan, I reread your use case and now I understand why you need this to happen so quickly. In this case, absolutely, I see the logic for checking for returns on the arrival of a new participant.

RE participant statuses via the Prolific API, yes, this is accessible via the Submissions route: https://docs.prolific.com/docs/api-docs/public/#tag/Submissions

**Comment by polvanrijn** [comment]
@thomasmorgan, I mean to periodically send an API request to the recruiter (Prolific or mturk). I just implemented this for another recruitment service (called Lucid) for Psynet. If you like I can explain how I did it here. But maybe it's not relevant for your use-case.

**Comment by thomasmorgan** [comment]
@pmcharrison @polvanrijn 
OK, this is great, thanks. Yes, I think this is the solution then. My only lingering concern is that checking all participant statuses might be slow for large experiments, but there's usually a good while between participant arrival and node creation, so hopefully its not too heavy.

@jessesnyder
Does this sound like a viable solution to you?

This is giving me flashbacks to trying to solve the missing notifications issue for MTurk! Glad we have better solutions available than what I was able to come up with back then.

**Comment by pmcharrison** [comment]
I think fortunately you can list all participant statuses within 1 API call, so it should be scalable: https://docs.prolific.com/docs/api-docs/public/#tag/Submissions/operation/GetSubmissions

**Comment by polvanrijn** [comment]
Yes, I can confirm this is available for all recruiters I have worked with (Prolific, mturk and Lucid).

**Comment by pmcharrison** [comment]
A further advantage of developing functionality via this API call is that it is very easy to test. In contrast, it's hard to develop a notification listener without real participants, especially in Prolific which doesn't have a sandbox like MTurk

**Comment by jessesnyder** [comment]
Some questions

1. I'm a little frightened by the prospect of making an API call every time a participant joins. This call will need to block, since we're using the result to determine whether to admit the prospective participant, correct? (I'm remembering Griduniverse experiments where very large numbers of people were attempting to join more-or-less simultaneously.) If none of you are worried about this then maybe it's probably not a real issue.
2. It seems like we'll still need some combination of a clock-triggered task and/or an event subscription, since at some point there will no longer be new participants joining to trigger a check, and we will want to continue to check for returned or abandoned participants, right?

**Comment by pmcharrison** [comment]
These solutions sounds quite complicated... Regarding this from @jessesnyder :

> I'm a little frightened by the prospect of making an API call every time a participant joins. This call will need to block, since we're using the result to determine whether to admit the prospective participant, correct? 

No, I don't think we are allowed to do that. Once a participant arrives from Prolific we have to let them take the experiment. I thought that instead @thomasmorgan would just take the opportunity to run some logic on the networks to make sure that the returned participant's nodes are failed, making room for the new participant. But that can be done asynchronously, no? As long as the experiment keeps the participant busy with something else for a few seconds before they reach the network code?

**Comment by jessesnyder** [comment]
If we consistently update participant statuses as we add each participant (even if this is done in an async worker, so effectively **after** we add them) the status in the DB will be at or near 100% accurate each time a new participant joins. Is this a close-enough reflection of Prolific? (Agreeing with @pmcharrison here)

**Comment by thomasmorgan** [comment]
@pmcharrison @jessesnyder 
Yes, you are both correct. Peter is right that we are doing this validation not to accept or reject participants, but rather to figure out where to place participants who arrive unexpectedly. I guess it is possible that for some experiments there is no where to put them, but then it is just a case of over-recruitment and so they are accepted regardless, jsut sent straight to debrief.

Jesse is correct that this will keep dallinger and prolific pretty much in sync, and given that this specific issue only arises when participants arrive, making that the trigger of syncing should make everything run smoothly. Once we have the sycning function in place, we might want to run it every 30 seconds anyway, just to be safe, but it shouldn't be important unless there are other issues we need to account for.

**Comment by pmcharrison** [comment]
I have been using the `scheduled_task` decorator a lot, it works great! Not for this purpose though. I've only used it for relatively fast processes and I'm not sure about doing API calls there, because a slow API response could disrupt other scheduled tasks (since I think is just a single process)? But you can use the scheduled task to queue worker processes.

I do like the idea of having the scheduled task running in addition to any participant triggers, it seems a good way of offering reassurance that the current status is up to date.

Unfortunately the sketches both gave 404s for me.

**Comment by jessesnyder** [comment]
Thanks, @pmcharrison. Yes, you're right about the single process blocking unless you pass off to a worker. Sorry about the links! These will be more reliable:

[Version 1](https://github.com/Dallinger/Dallinger/blob/issues/5725-prolific-duration-exceeded/docs/diagrams/participant_status_sync_v1.md), where the experiment is primarily responsible for responding to the scheduled "pings"
[Version 2](https://github.com/Dallinger/Dallinger/blob/issues/5725-prolific-duration-exceeded/docs/diagrams/participant_status_sync_v2.md), where the recruiter and dallinger do a lot more of the work

**Comment by jacobyn** [comment]
I think that the best solution is to implement the polling option. Namely that we have  a route that update the information via an API call to prolific and we set a background process to update this information. We can additionally call this route every time a participant submitting to make sure it is also updated then. I think this solution is easy to implement and address all of the issues. If the experimented needs this information to be updated in another place in the experiment they can call this route as well at any point they desire.

**Comment by jessesnyder** [comment]
Is it OK to have recruiters initiate the polling (v2 above), or would you prefer to have the experiment instance get the time-based notifications and be responsible for calling the recruiter to find discrepancies, and then dealing with the results (v1 above)?

**Comment by pmcharrison** [comment]
I think best to put this logic in the Recruiter, yes.

**Comment by pmcharrison** [comment]
(so v2)

**Comment by thomasmorgan** [comment]
I agree that we should put this stuff in the recruiter, especially as details are going to be recruiter specific. However, when a participant arrives, won't it be the experiment that detects this and so requests that participant statuses are updated? If so, this means not all the code can be in the recruiter. Although maybe the recruiter does some participant onboarding too?

**Comment by pmcharrison** [comment]
Yes, we can have the Experiment call `ping_recruiters` during participant onboarding. We do need to be a bit careful about race conditions (if the Experiment's call overlaps with the background task) but this can be done with careful SQLAlchemy I guess

**Comment by jessesnyder** [comment]
Agreed that with v2, the pathway to updating existing participants when a new one joins is through `ping_recruiters` (the experiment can just enqueue a call to `ping_recruiters` with no arguments, just like the clock does).

I think this model will help avoid race condition problems, since each execution of `ping_recruiters` will fetch the latest participant statuses from the recruitment service, and I assume the queue gets processed FIFO, so it will always be more current info overwriting older info(?) Tell me what I'm missing. :-)

**Comment by pmcharrison** [comment]
@thomasmorgan SQLAlchemy only locks rows if you add `.with_for_update()` to your query, normally it doesn't matter but sometimes it does and it can be painful!

> I think there remains a race condition though: If a ppt arrives and pings the recruiters, the check will get queued, but if the participant tries to make a node before the check completes we can get issues. I'm not sure how severe this is, as the check should take <1s and participants have to click through stuff to get to node creation. But if an experiment tried to make a node as soon as the ppt was created, we could have issues, no?

Maybe this is best addressed via experiment design? e.g. give the participant instructions before creating nodes?

**Comment by jessesnyder** [comment]
Hmm.. it's starting to sound like we're talking about **blocking** participant creation while we wait for an update from the recruiter. I was proposing enqueuing an asynchronous request for status updates when a participant joins. This would be a strategy for generally keeping status up to date, rather than guaranteeing that all status is up to date before creating a participant (or assigning a participant to a node). Am I misinterpreting?

**Comment by thomasmorgan** [comment]
We don't need to block _participant_ creation per se. We just need to make sure all participant statuses are up to date before the incoming participant tries to make a _node_ and join a network. That said, I guess, blocking the participant creation while the status updates occur is one solution.

**Comment by jessesnyder** [comment]
This sounds really complicated. Why isn't this an issue with the MTurk recruiter? We only get asynchronous push notifications via AWS SNS for this service. (Or maybe it's always been an issue with MTurk also?)

**Comment by thomasmorgan** [comment]
Its not a problem with MTurk because MTurk sends returned/abadoned notifications before the replacement participant arrives, so the database is always up to date.

In the past, these notifications from AWS did fail to arrive sometimes, and it caused huge issues.

When a participant returns or abandons, we need to know so we can delete their data before their replacement turns up. AWS tells us via a notification, Prolific does not. So for prolific we need to find a way to do this ourselves, and critically we need to do it in the small window before the replacement participant tries to join a network.

Does this make sense?

**Comment by jessesnyder** [comment]
Makes sense. It seems like its the extremely eager participant replacement in Prolific that's the core issue.

An idea:
Instead of having the `create_participant` route in Dallinger be directly responsible for tallying up active participants, the route could instead delegate this task to the recruiters with a blocking query. MTurk and built-in recruiters (HotAir and friends) could perform a query of the local DB, since that's a trusted source of truth for them, but Prolific could call out to the mother ship.

**Comment by thomasmorgan** [comment]
Yes, it is amazing how quickly prolific can pipe in a replacement.

Having the tallying code in the recruiter is a good idea as it will vary by recruiter, and yeah, having a blocking query for prolific sounds good.

**Comment by jessesnyder** [comment]
@pmcharrison @polvanrijn Thoughts on this idea? (starting [here](https://github.com/Dallinger/Dallinger/issues/5725#issuecomment-1965174194))

**Comment by pmcharrison** [comment]
Do you forsee a problem if many participants hit the same blocking route at the same time?

**Comment by jessesnyder** [comment]
Seems quite possible. (Off to the dentist... Maybe brilliance will strike while in the chair 🦷  -> 💡)

**Comment by pmcharrison** [comment]
Good luck at the dentist!

Yes, to me it seems undesirable to make one of our own HTTP requests dependent on a blocking API call. I think it's better practice to serve such requests as soon as possible... Otherwise server performance could slow significantly.

I think it would be safest for now to trigger the check asynchronously and make it the experimenter's obligation to give the participant something to do before creating nodes etc.

**Comment by jessesnyder** [comment]
> make it the experimenter's obligation to give the participant something to do before creating nodes etc

@thomasmorgan Can make do within these constraints?

**Comment by thomasmorgan** [comment]
@jessesnyder yes, this works for me.

I actually think, even if this does become an issue, its actually going to self correct. If a participant tries to make a node before the database is updated, they'll be kicked out, but this will cause another replacement to arrive, by which time the database will have updated and the new replacement can be put in fine.

**Comment by pmcharrison** [comment]
This is now working, at least the part where a background process polls Prolific and updates participant statuses. @thomasmorgan it should be quite straightforward now to create additional calls to this 'recruiter checks' function when you like, e.g. on node creation or similar.

**Comment by thomasmorgan** [comment]
@pmcharrison Perfect, thank for letting me know!

**Comment by thomasmorgan** [comment]
@pmcharrison @jessesnyder Hi both, I'm trying to implement this fix in my current study. I need to force the db to sync with prolific whenever a new participant arrives. I am doing this by overwriting the `create_participant` function as follows:
`    def create_participant(
        self,
        worker_id,
        hit_id,
        assignment_id,
        mode,
        recruiter_name=None,
        fingerprint_hash=None,
        entry_information=None,
    ):
        """ Needed for prolific. When a new participant arrives we force a query of prolifics
        participant statuses to check if anyone has returned. """
        q = db.get_queue()
        q.enqueue(recruiters.run_status_check)
        participant = super(Epivigi, self).create_participant(
            self,
            worker_id,
            hit_id,
            assignment_id,
            mode,
            recruiter_name,
            fingerprint_hash,
            entry_information,
        )
        return participant`

Does this look right to you?

**Comment by jacobyn** [comment]
I think this should work now (sync already) so if you are using new version it should work already. Right?

**Comment by thomasmorgan** [comment]
I'm running on Dallinger v11.1.1 and it is not auto-catching and syncing participants who return. My understanding is that in these cases the experimenter needs to force a sync when necessary, which is what I'm trying to do.

That said, if there is a background task that loops this, it should catching things eventually, but I didn't see this happening. What does the user need to do to activate this background task? I notice I have `clock_on = false` in my config, could this be disabling it?

**Comment by jessesnyder** [comment]
Yes, it's definitely a clock/scheduled task that runs the automatic check:

```
@scheduler.scheduled_job("interval", minutes=0.5)
def async_recruiter_status_check():
    """Ask recruiters to check the status of their participants"""

    q = db.get_queue()
    q.enqueue(recruiters.run_status_check)

```

**Comment by thomasmorgan** [comment]
Ah okay, perfect, this is a very easy fix. For clarity, if I want to force a sync manually, all I need to do is import the db and then:
```
q = db.get_queue()
       q.enqueue(recruiters.run_status_check)
```
Right?

**Comment by jessesnyder** [comment]
That looks right to me!

**Comment by pmcharrison** [comment]
Yes, or if you want the check to block (so that by the time you reach the next line you know that the checks have been successfully completed) then just `recruiters.run_status_check()` should do it.

**Comment by thomasmorgan** [comment]
Perfect, thanks both.
